File: DenoisingFilters.py
import os
import cv2
import numpy as np
from scipy.ndimage import minimum_filter, maximum_filter
from scipy.signal import wiener

NOISY_ROOT = "noisy"
DENOISED_ROOT = "denoised"

os.makedirs(DENOISED_ROOT, exist_ok=True)

def adaptive_wiener_color(img):
    if len(img.shape) == 3:
        channels = []
        for c in range(3):
            ch = wiener(img[:, :, c], (5, 5))
            ch = np.clip(ch, 0, 255).astype(np.uint8)
            channels.append(ch)
        return cv2.merge(channels)
    else:
        out = wiener(img, (5, 5))
        return np.clip(out, 0, 255).astype(np.uint8)

def apply_filters(img):
    return {
        "gaussian_filter": cv2.GaussianBlur(img, (5, 5), 0),
        "median_filter": cv2.medianBlur(img, 5),
        "min_filter": minimum_filter(img, size=3).astype(np.uint8),
        "max_filter": maximum_filter(img, size=3).astype(np.uint8),
        "adaptive_filter": adaptive_wiener_color(img)
    }

noise_types = [d for d in os.listdir(NOISY_ROOT) if os.path.isdir(os.path.join(NOISY_ROOT, d))]

for noise_type in noise_types:
    in_dir = os.path.join(NOISY_ROOT, noise_type)
    files = [f for f in os.listdir(in_dir) if f.lower().endswith((".jpg", ".jpeg", ".png", ".bmp"))]

    print(f"\nProcessing noise type: {noise_type} | {len(files)} images")

    for idx, file in enumerate(files, start=1):
        img_path = os.path.join(in_dir, file)
        img = cv2.imread(img_path)

        if img is None:
            print(f"Skipped unreadable image: {img_path}")
            continue

        filtered = apply_filters(img)

        for name, fimg in filtered.items():
            out_dir = os.path.join(DENOISED_ROOT, name, noise_type)
            os.makedirs(out_dir, exist_ok=True)

            out_path = os.path.join(out_dir, file)

            if os.path.exists(out_path):
                continue

            cv2.imwrite(out_path, fimg)

        if idx % 50 == 0 or idx == len(files):
            print(f"{noise_type}: {idx}/{len(files)} done")

print("\nAll denoising finished.")

File: Denoising_Filters.m
inputRoot = 'noisyMatlab';
outputRoot = 'denoisedMatlab';

if ~exist(outputRoot, 'dir')
    mkdir(outputRoot);
end

filterNames = {'gaussian_filter', 'median_filter', 'min_filter', 'max_filter', 'adaptive_filter'};

for i = 1:length(filterNames)
    outDir = fullfile(outputRoot, filterNames{i});
    if ~exist(outDir, 'dir')
        mkdir(outDir);
    end
end

noiseFolders = dir(inputRoot);
noiseFolders = noiseFolders([noiseFolders.isdir]);
noiseFolders = noiseFolders(~ismember({noiseFolders.name}, {'.', '..'}));

for n = 1:length(noiseFolders)
    noiseType = noiseFolders(n).name;
    inDir = fullfile(inputRoot, noiseType);

    files = dir(fullfile(inDir, '*.*'));
    files = files(~[files.isdir]);

    for i = 1:length(files)
        fileName = files(i).name;
        imgPath = fullfile(inDir, fileName);
        img = imread(imgPath);

        if size(img,3) == 3
            gray = rgb2gray(img);
        else
            gray = img;
        end

        % 1. Gaussian filter
        g = imgaussfilt(img, 2);
        imwrite(g, fullfile(outputRoot, 'gaussian_filter', [noiseType '_' fileName]));

        % 2. Median filter
        if size(img,3) == 3
            m = img;
            for c = 1:3
                m(:,:,c) = medfilt2(img(:,:,c), [3 3]);
            end
        else
            m = medfilt2(img, [3 3]);
        end
        imwrite(m, fullfile(outputRoot, 'median_filter', [noiseType '_' fileName]));

        % 3. Min filter
        if size(img,3) == 3
            minf = img;
            for c = 1:3
                minf(:,:,c) = ordfilt2(img(:,:,c), 1, true(3));
            end
        else
            minf = ordfilt2(img, 1, true(3));
        end
        imwrite(minf, fullfile(outputRoot, 'min_filter', [noiseType '_' fileName]));

        % 4. Max filter
        if size(img,3) == 3
            maxf = img;
            for c = 1:3
                maxf(:,:,c) = ordfilt2(img(:,:,c), 9, true(3));
            end
        else
            maxf = ordfilt2(img, 9, true(3));
        end
        imwrite(maxf, fullfile(outputRoot, 'max_filter', [noiseType '_' fileName]));

        % 5. Adaptive Wiener filter
        if size(img,3) == 3
            af = img;
            for c = 1:3
                af(:,:,c) = uint8(wiener2(img(:,:,c), [5 5]));
            end
        else
            af = uint8(wiener2(img, [5 5]));
        end
        imwrite(af, fullfile(outputRoot, 'adaptive_filter', [noiseType '_' fileName]));
    end
end

disp('Denoised images saved successfully.');